[primer] Create a class for easier comparison of pylint results

Pierre-Sassoulas opened the pull request:
## Type of Changes

<!-- Leave the corresponding lines for the applicable type of change: -->

|     | Type                   |
| --- | ---------------------- |
| ✓   | :hammer: Refactoring   |

## Description

Follow up to #6973

DanielNoord commented on pylint/testutils/_primer/comparator.py:
```diff
@@ -0,0 +1,41 @@
+from collections.abc import Generator
+from pathlib import Path
+
+from pylint.testutils._primer.primer_command import Messages, PackageMessages
```
Could you do a follow up PR that moves some of these to `typing.py` file?

Pierre-Sassoulas replied:
I'm starting to think we should use the existing class for this, I'll give it a try.

Pierre-Sassoulas replied:
Resolving as we merged #7000

DanielNoord commented on pylint/testutils/_primer/primer_command.py:
```diff
@@ -11,7 +11,8 @@
 from pylint.testutils._primer import PackageToLint
 
-PackageMessages = Dict[str, List[Dict[str, Union[str, int]]]]
+Messages = List[Dict[str, Union[str, int]]]
```
Never mind my earlier comment. Perhaps you extract these now.

DanielNoord commented on pylint/testutils/_primer/comparator.py:
```diff
@@ -0,0 +1,41 @@
+from pylint.testutils._primer.primer_command import Messages, PackageMessages
+
+
+class Comparator:
```
Could you rename this to `PylintComparator`? At some point I want to make an `AstroidComparator` 😄 

If you're up for it perhaps we could even create an abstract class?

DanielNoord replied:
No, using the `primer` on `astroid/main` to catch crashes there before we merge them. That would require a different `Comparator` I think.

DanielNoord replied:
Yeah, but I thinking of maybe running the `pylint` test suite on some version on `astroid` `main` or something like that. The overall goal is to remove the need for `pylint-tested` label and have all PRs get that automatically.

Pierre-Sassoulas replied:
Yeah sure I understood that :) It make sense to test astroid more in the astroid CI, because pylint's CI is already crowded so I think it's a really good idea too.

Pierre-Sassoulas committed:
Extract Comparator class from CompareCommand

Move JSON loading, batching, and cross-referencing logic into a
dedicated Comparator class. CompareCommand.run() now delegates to
the Comparator and iterates over it to build the comment.

DanielNoord commented:
@Pierre-Sassoulas I would be okay with rebasing this and not doing the rename if that means we can merge this easily :)

Pierre-Sassoulas commented:
I think there was a concurrent change that forced to redo everything or at least a lot. (Which is why it's stuck not the rename). Might aknowledge the loss and redo from scratch later instead of trying to rebase then getting demoralised :)  Also there was a gazillion thing on my todo list before that (too-few-logging and always suggest for no-member) but we now have a cache problem to fix first :D...

DanielNoord reviewed (commented):
Cool! Some small nits, but I'll stay on top of reviewing this so we can merge this and not let this get stale again.

Pierre-Sassoulas committed:
Extend duplicate-code disable to cover tempdir helper

Move the ``# pylint: enable=duplicate-code`` comment after the
``tempdir`` context manager so it is covered by the existing
suppression scope.

DanielNoord commented:
@Pierre-Sassoulas Feels like a rebase went wrong here? There are some unrelated dependency updates in this PR.

DanielNoord reviewed (commented):
Except for the unrelated changes this LGTM!

@Pierre-Sassoulas Can you rebase again? I'll approve and merge afterwards!
